File: backend/services/userService.py
from repositories import userRepository

class UserService:
    def __init__(self):
        self.userRepo = userRepository.UserRepository()

    def get_users(self):
        userJson = []

        for user in self.userRepo.get_users():
            userJson.append(user.to_json())

        return userJson

    def add_user(self, firstname, lastname, role, imageUrl=None):
        from entities.user import User
        user = User(firstName=firstname, lastName=lastname, role=role, imageUrl=imageUrl)
        return self.userRepo.add_user(user).to_json()

    def update_user(self, user_id, firstname, lastname, role, imageUrl=None):
        user = self.userRepo.get_user_by_id(user_id)

        if not user:
            return None

        user = self.userRepo.update_user(user_id, firstname, lastname, role, imageUrl)

        if not user:
            return None

        return user.to_json()

    def delete_user(self, user_id):
        return self.userRepo.delete_user(user_id)

    def get_users_by_role(self, role):
        userJson = []

        for user in self.userRepo.get_users_by_role(role):
            userJson.append(user.to_json())

        return userJson

File: frontend/src/app/services/user.service.ts
import { Injectable } from "@angular/core";
import { HttpClient } from "@angular/common/http";
import { Observable } from "rxjs";
import { IUser } from "../model/user.interface";
import { IResponse } from "../model/response.interface";

@Injectable({
  providedIn: "root",
})
export class UserService {
  BASE_URL: string = "http://localhost:5000/api/users";

  constructor(private http: HttpClient) {}

  getAll(): Observable<IResponse<IUser[]>> {
    return this.http.get<IResponse<IUser[]>>(this.BASE_URL);
  }

  getAllClients() :Observable<IResponse<IUser[]>> {
    return this.http.get<IResponse<IUser[]>>(this.BASE_URL + "/roles/0");
  }

  addUser(
    firstname: string,
    lastname: string,
    role: number,
    imageUrl: string | null,
  ): Observable<IResponse<IUser>> {
    const payload = {
      firstname: firstname,
      lastname: lastname,
      role: role,
      imageUrl: imageUrl,
    };
    return this.http.post<IResponse<IUser>>(this.BASE_URL, payload);
  }

  updateUser(
    userId: number,
    firstname: string,
    lastname: string,
    role: number,
    imageUrl: string | null,
  ): Observable<IResponse<IUser>> {
    const payload = {
      firstname: firstname,
      lastname: lastname,
      role: role,
      imageUrl: imageUrl,
    };
    return this.http.put<IResponse<IUser>>(
      `${this.BASE_URL}/${userId}`,
      payload
    );
  }

  deleteUser(userId: number): Observable<IResponse<null>> {
    return this.http.delete<IResponse<null>>(`${this.BASE_URL}/${userId}`);
  }
}
